pkg/math/sample/sample.go

```diff
@@ -200,42 +201,43 @@ const maxPrimeIterations = 100_000
 // ErrMaxPrimeIterations is the error we return when we fail to generate a prime.
 var ErrMaxPrimeIterations = fmt.Errorf("sample: failed to generate prime after %d iterations", maxPrimeIterations)
 
-// BlumPrime returns a safe prime p of size params.BitsBlumPrime.
-//
-// This means that q := (p - 1) / 2 is also a prime number.
-//
-// This implies that p = 3 mod 4, otherwise q wouldn't be prime.
-func BlumPrime(rand io.Reader) *safenum.Nat {
-	// TODO be more flexible on the number of bits in P, Q to avoid square root attack
-	one := new(big.Int).SetUint64(1)
-	for i := 0; i < maxPrimeIterations; i++ {
-		p, err := potentialSafePrime(rand, params.BitsBlumPrime)
-		if err != nil {
-			continue
-		}
-		// For p to be safe, we need q := (p - 1) / 2 to also be prime
-		q := new(big.Int).Sub(p, one)
-		q.Rsh(q, 1)
-		// p is likely to be prime already, so let's first do the other check,
-		// which is more likely to fail.
-		if !q.ProbablyPrime(blumPrimalityIterations) {
-			continue
-		}
-		// We've only done light checks on p so far, so now we need to make
-		// sure that it passes the extensive ones
-		if !p.ProbablyPrime(blumPrimalityIterations) {
-			continue
-		}
-		return new(safenum.Nat).SetBig(p, params.BitsBlumPrime)
+var one = new(big.Int).SetUint64(1)
+
+func tryBlumPrime(rand io.Reader) *safenum.Nat {
+	p, err := potentialSafePrime(rand, params.BitsBlumPrime)
+	if err != nil {
+		return nil
+	}
+	// For p to be safe, we need q := (p - 1) / 2 to also be prime
+	q := new(big.Int).Sub(p, one)
```

Comment:
I just realized we don't actually need to subtract one if we right shift, since p is odd.

Reply:
Oh yeah, I'll fix this up after merging!